File: Problem202_HappyNumber.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long sumOfSquare(int n) {
        long long res=0;
        while (n>0) {
            int temp = n%10;
            n/=10;
            res += temp*temp;
        }
        return res;
    }
    bool isHappy(int n) {
        long long temp = sumOfSquare(n);
        if (temp==1) return true;
        unordered_map<int, int> check;
        while (temp != n) {
            cout << temp << " ";
            if (check[temp]) return false;
            check[temp]=1;
            if (temp==1) return true;
            temp = sumOfSquare(temp);
        }
        cout << endl;
        return false;
    }
};

int main() {
    const int n=4;
    Solution solve;
    bool isHap = solve.isHappy(n);
    cout << isHap << INT_MAX;
}

File: Problem202_HappyNumber.py
class Solution:
    def sumOfSquares(self, n):
        res = 0
        while n!=0:
            res+=pow(n%10, 2)
            n//=10
        return res
    def isHappy(self, n: int) -> bool:
        temp = self.sumOfSquares(n)
        if temp==1: return True
        check = {}
        while temp!=n:
            print(temp)
            if temp not in check.keys(): check[temp]=0
            if check[temp]: return False
            if temp==1: return True
            check[temp]=1
            temp = self.sumOfSquares(temp) 
        return False

if __name__ == "__main__":
    solve = Solution()
    n = 2147483647
    print(solve.isHappy(n))